staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go

```diff
@@ -67,65 +71,106 @@ type kmsPluginProbe struct {
 	l            *sync.Mutex
 }
 
+type kmsv2PluginProbe struct {
+	name string
+	ttl  time.Duration
+	envelopekmsv2.Service
+	lastResponse *kmsPluginHealthzResponse
+	l            *sync.Mutex
+}
+
 func (h *kmsPluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
 	return healthz.NamedCheck(fmt.Sprintf("kms-provider-%d", idx), func(r *http.Request) error {
 		return h.Check()
 	})
 }
 
+func (p *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
+	return healthz.NamedCheck(fmt.Sprintf("kmsv2-provider-%d", idx), func(r *http.Request) error {
+		return p.Check()
+	})
+}
+
 // GetKMSPluginHealthzCheckers extracts KMSPluginProbes from the EncryptionConfig.
 func GetKMSPluginHealthzCheckers(filepath string) ([]healthz.HealthChecker, error) {
 	f, err := os.Open(filepath)
 	if err != nil {
 		return nil, fmt.Errorf("error opening encryption provider configuration file %q: %v", filepath, err)
 	}
 	defer f.Close()
+
 	var result []healthz.HealthChecker
-	probes, err := getKMSPluginProbes(f)
+	probes, probesKMSv2, err := getKMSPluginProbes(f)
 	if err != nil {
 		return nil, err
 	}
-
 	for i, p := range probes {
 		probe := p
 		result = append(result, probe.toHealthzCheck(i))
 	}
+	for i, p := range probesKMSv2 {
+		probe := p
+		result = append(result, probe.toHealthzCheck(i))
+	}
+
 	return result, nil
 }
 
-func getKMSPluginProbes(reader io.Reader) ([]*kmsPluginProbe, error) {
+func getKMSPluginProbes(reader io.Reader) ([]*kmsPluginProbe, []*kmsv2PluginProbe, error) {
```

Comment:
Instead of differentiating between v1 and v2, can we just return and interface here and append v2 plugins after v1 plugins if the feature gate is enabled?

Reply:
Done!